e2e/pos_test.go

```diff
@@ -0,0 +1,617 @@
+			Input:    framework.MethodSig("unstake"),
+		}, senderKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		currentNonce++
+
+		return signedTx
+	}
+
+	zeroEth := framework.EthToWei(0)
+	for i := 0; i < numTransactions; i++ {
+		var msg *txpoolOp.AddTxnReq
+		unstakeTxn := generateTx()
+
+		msg = &txpoolOp.AddTxnReq{
+			Raw: &any.Any{
+				Value: unstakeTxn.MarshalRLP(),
+			},
+			From: types.ZeroAddress.String(),
+		}
+
+		_, addErr := clt.AddTxn(context.Background(), msg)
+		if addErr != nil {
+			t.Fatalf("Unable to add txn, %v", addErr)
+		}
+	}
+
+	// Set up the blockchain listener to catch the added block event
+	blockNum := waitForBlock(t, srv, 1, 0)
+
+	block, blockErr := client.Eth().GetBlockByNumber(web3.BlockNumber(blockNum), true)
+	if blockErr != nil {
+		t.Fatalf("Unable to fetch block")
+	}
+
+	// Find how much the account paid for all the transactions in this block
+	paidFee := big.NewInt(0).Mul(bigGasPrice, big.NewInt(int64(block.GasUsed)))
+
+	// Check the balances
+	actualAccountBalance := framework.GetAccountBalance(senderAddr, client, t)
+	actualStakingSCBalance, fetchError = framework.GetStakedAmount(senderAddr, client)
+	if fetchError != nil {
+		t.Fatalf("Unable to fetch staking SC balance, %v", fetchError)
+	}
+
+	// Make sure the balances match up
+
+	// expBalance = previousAccountBalance + stakeRefund - block fees
+	expBalance := big.NewInt(0).Sub(big.NewInt(0).Add(previousAccountBalance, bigDefaultStakedBalance), paidFee)
+
+	assert.Equalf(t,
+		expBalance.String(),
+		actualAccountBalance.String(),
+		"Account balance mismatch after unstake exploit",
+	)
+
+	assert.Equalf(t,
+		zeroEth.String(),
+		actualStakingSCBalance.String(),
+		"Staked address balance mismatch after unstake exploit",
+	)
+}
+
+// generateStakingAddresses is a helper method for generating dummy staking addresses
+func generateStakingAddresses(numAddresses int) []types.Address {
+	result := make([]types.Address, numAddresses)
+
+	for i := 0; i < numAddresses; i++ {
+		result[i] = types.StringToAddress(strconv.Itoa(i + 100))
+	}
+
+	return result
+}
+
+// Test scenario:
+// User has 10 ETH staked and a balance of 100 ETH
+// Unstake -> Stake 1 ETH -> Unstake -> Stake 1 ETH...
+// The code below tests (numTransactions / 2) cycles of Unstake -> Stake 1 ETH
+// Expected result for tests: Staked: 1 ETH; Balance: ~119 ETH
+func TestPoS_StakeUnstakeExploit(t *testing.T) {
+	// Predefined values
+	var blockGasLimit uint64 = 5000000000
+	stakingContractAddr := staking.AddrStakingContract
+	bigDefaultStakedBalance := getBigDefaultStakedBalance(t)
+	defaultBalance := framework.EthToWei(100)
+	bigGasPrice := big.NewInt(framework.DefaultGasPrice)
+
+	senderKey, senderAddr := tests.GenerateKeyAndAddr(t)
+	numDummyStakers := 100
+
+	devInterval := 5 // s
+
+	// Set up the test server
+	srvs := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
```

Comment:
More validators please. At least 2.

Reply:
This test actually has 101 stakers:
https://github.com/0xPolygon/polygon-sdk/blob/eb35d49a64ab3c8d044e3333c09e68cab7a6443e/e2e/pos_test.go#L380